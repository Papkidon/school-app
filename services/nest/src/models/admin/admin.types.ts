import { z } from 'zod';

export const adminCreateSchema = z.object({
  data: z.object({
    username: z.string().min(2).max(20),
    password: z.string().min(8).max(20),
  }),
});

export const adminUpdateSchema = adminCreateSchema.deepPartial();

export type adminType = {
  data: {
    username: string;
    password: string;
  };
};
