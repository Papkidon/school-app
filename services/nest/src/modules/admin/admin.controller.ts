import {
  Body,
  Controller,
  Delete,
  Get,
  HttpException,
  HttpStatus,
  Param,
  Patch,
  Post,
} from '@nestjs/common';
import {
  adminCreateSchema,
  adminUpdateSchema,
} from '../../models/admin/admin.types';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { Admin } from '@prisma/client';
import { adminType } from '../../models/admin/admin.types';
import { DeepPartial, idSchema } from '../../models/utility.types';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import { AdminService } from './admin.service';

@Controller('api/v1/admin')
export class AdminController {
  constructor(private readonly adminService: AdminService) {}

  @Get('')
  async findAll(): Promise<Admin[]> {
    try {
      const result = await this.adminService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No admin data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(adminCreateSchema)) createAdminDto: adminType,
  ): Promise<Admin> {
    try {
      const userExists = await this.adminService.findByUsername(
        createAdminDto.data.username,
      );

      if (userExists) {
        throw new HttpException(
          'User with given username already exists',
          HttpStatus.CONFLICT,
        );
      }

      return await this.adminService.create(createAdminDto);
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Admin> {
    try {
      const result = await this.adminService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(adminUpdateSchema))
    body: DeepPartial<adminType>,
  ): Promise<Admin> {
    try {
      const userExists = await this.adminService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'Admin with given id does not exist',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.adminService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Admin> {
    try {
      const userExists = await this.adminService.findById(id);

      if (!userExists) {
        throw new HttpException(
          'Admin with given id does not exist',
          HttpStatus.NOT_FOUND,
        );
      }

      return await this.adminService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
