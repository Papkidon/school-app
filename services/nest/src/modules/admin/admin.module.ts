import { Module } from '@nestjs/common';
import { PrismaService } from 'nestjs-prisma';
import { AdminController } from './admin.controller';
import { AdminRepository } from './admin.repository';
import { AdminService } from './admin.service';

@Module({
  controllers: [AdminController],
  providers: [AdminService, AdminRepository, PrismaService],
})
export class AdminModule {}
