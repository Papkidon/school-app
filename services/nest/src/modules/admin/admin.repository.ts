import { Injectable } from '@nestjs/common';
import { PrismaService as PrismaService } from 'nestjs-prisma';
import { adminType } from '../../models/admin/admin.types';
import { DeepPartial } from '../../models/utility.types';
import { Admin } from '@prisma/client';

@Injectable()
export class AdminRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<Admin[]> {
    return await this.prisma.admin.findMany();
  }

  async findById(id: string): Promise<Admin> {
    return await this.prisma.admin.findUnique({ where: { id: id } });
  }

  async create(body: any): Promise<Admin> {
    return await this.prisma.admin.create(body);
  }

  async delete(id: string): Promise<Admin> {
    return await this.prisma.admin.delete({ where: { id: id } });
  }

  async update(id: string, body: DeepPartial<adminType>): Promise<Admin> {
    return await this.prisma.admin.update({
      where: { id: id },
      data: { ...body.data },
    });
  }
}
