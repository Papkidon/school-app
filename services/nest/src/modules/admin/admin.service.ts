import { Injectable } from '@nestjs/common';
import { AdminRepository } from './admin.repository';
import { adminType } from '../../models/admin/admin.types';
import { DeepPartial } from '../../models/utility.types';
import { Admin } from '@prisma/client';

@Injectable()
export class AdminService {
  constructor(private adminRepository: AdminRepository) {}
  async findAll(): Promise<Admin[]> {
    return await this.adminRepository.findAll();
  }

  async findById(id: string): Promise<Admin> {
    return await this.adminRepository.findById(id);
  }

  async create(body: any): Promise<Admin> {
    return await this.adminRepository.create(body);
  }

  async delete(id: string): Promise<Admin> {
    return await this.adminRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<adminType>): Promise<Admin> {
    return await this.adminRepository.update(id, body);
  }
}
