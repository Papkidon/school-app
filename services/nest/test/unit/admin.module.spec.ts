import { Admin } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import { AdminController } from '../../src/modules/admin/admin.controller';
import { AdminRepository } from '../../src/modules/admin/admin.repository';
import { AdminService } from '../../src/modules/admin/admin.service';
import {
  dummyAdminCreate,
  dummyAdminId,
  dummyAdminUpdate,
} from './utils/dummy.data';

describe('UserController', () => {
  let adminController: AdminController;
  let adminService: AdminService;
  let adminRepository: AdminRepository;
  let prismaService: PrismaService;

  beforeEach(() => {
    prismaService = new PrismaService();
    adminRepository = new AdminRepository(prismaService);
    adminService = new AdminService(adminRepository);
    adminController = new AdminController(adminService);
  });

  afterAll(() => {
    jest.useRealTimers();
    jest.resetAllMocks();
  });

  describe('findAll', () => {
    it('should return an array of all admins', async () => {
      const result = ['admin1', 'admin2'] as unknown as Admin[];
      jest
        .spyOn(adminRepository, 'findAll')
        .mockImplementation(async () => result);

      expect(await adminController.findAll()).toMatchSnapshot();
    });
  });

  describe('findUnique', () => {
    it('should return an admin with given id', async () => {
      const result = 'dummyFoundByIdAdmin' as unknown as Admin;
      jest
        .spyOn(adminRepository, 'findById')
        .mockImplementation(async () => result);

      expect(await adminController.findById(dummyAdminId)).toMatchSnapshot();
    });
  });

  describe('create', () => {
    it('should create a new admin', async () => {
      const dummyAdminNotFound = '' as unknown as Admin;
      const result = 'dummyCreatedAdmin' as unknown as Admin;
      jest
        .spyOn(adminRepository, 'create')
        .mockImplementation(async () => result);
      jest
        .spyOn(adminRepository, 'findByUsername')
        .mockImplementation(async () => dummyAdminNotFound);

      expect(await adminController.create(dummyAdminCreate)).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('should update an admin with given id', async () => {
      const dummyFoundAdminById = 'dummyAdminExists' as unknown as Admin;
      const result = 'dummyUpdatedAdmin' as unknown as Admin;
      jest
        .spyOn(adminRepository, 'update')
        .mockImplementation(async () => result);
      jest
        .spyOn(adminRepository, 'findById')
        .mockImplementation(async () => dummyFoundAdminById);

      expect(
        await adminController.update(dummyAdminId, dummyAdminUpdate),
      ).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('should delete an admin with given id', async () => {
      const dummyAdminFoundById = 'dummyAdminFoundById' as unknown as Admin;
      const result = 'dummyDeletedAdmin' as unknown as Admin;
      jest
        .spyOn(adminRepository, 'delete')
        .mockImplementation(async () => result);
      jest
        .spyOn(adminRepository, 'findById')
        .mockImplementation(async () => dummyAdminFoundById);

      expect(await adminController.delete(dummyAdminId)).toMatchSnapshot();
    });
  });
});
