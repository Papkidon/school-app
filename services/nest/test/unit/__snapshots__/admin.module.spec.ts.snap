// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController create should create a new admin 1`] = `"dummyCreatedAdmin"`;

exports[`UserController delete should delete an admin with given id 1`] = `"dummyDeletedAdmin"`;

exports[`UserController findAll should return an array of all admins 1`] = `
[
  "admin1",
  "admin2",
]
`;

exports[`UserController findUnique should return an admin with given id 1`] = `"dummyFoundByIdAdmin"`;

exports[`UserController update should update an admin with given id 1`] = `"dummyUpdatedAdmin"`;
